File: implementations/rust/src/tsp.rs
use std::cmp::min;

use crate::espprc;

pub struct TSPData {
    pub n: usize,
    pub d: Box<[i32]>,
    pub aux: Box<[f64]>,
}

impl TSPData {
    pub fn d(&self, i: usize, j:usize) -> i32 { self.d[i * self.n + j] }

    pub fn aux(&self, i: usize, j:usize) -> f64 { self.aux[i * self.n + j] }
    
    pub fn setaux(&mut self, i: usize, j:usize, val:f64) {
        self.aux[i * self.n + j] = val;
    }
}

pub struct TSPSolution {
    pub nodes: Vec<usize>,
}

impl TSPSolution {
    pub fn two_opt(&mut self, data: &TSPData) -> i32 {
        let mut t = 0;
        while self.first_2e_improvement(&data) { t += 1; }
        return t;
    }
    
    fn first_2e_improvement(&mut self, data: &TSPData) -> bool {
        for (p1, n1) in self.nodes.windows(2).enumerate() {
            let (s1, t1) = (n1[0], n1[1]);
            for (p2, n2) in self.nodes.windows(2).enumerate().skip(p1 + 2) {
                let (s2, t2) = (n2[0], n2[1]);
                if data.d(s1, t1) + data.d(s2, t2) >
                    data.d(s1, s2) + data.d(t1, t2) {
                        for i in 0..((p2 - p1 + 1) / 2) {
                            self.nodes.swap(p1 + 1 + i, p2 - i);
                        }
                        return true;
                    }
            }
        }
        return false;
    }

    pub fn or_opt(&mut self, data: &TSPData) -> i32 {
        let mut t = 0;
        while self.first_or_improvement(&data) { t += 1; }
        return t;
    }

    fn or_delta(&self, data: &TSPData,
                ipred: usize, inode:usize,
                enode:usize, esucc:usize,
                pnode:usize, psucc:usize) -> i32 {
        return data.d(ipred, esucc) + data.d(pnode, inode)
            + data.d(enode, psucc) - data.d(pnode, psucc)
            - data.d(ipred, inode) - data.d(enode, esucc);
    }
    
    fn first_or_improvement(&mut self, data: &TSPData) -> bool {
        for (iminus, n1) in self.nodes.windows(2).enumerate() {
            let (ipred, inode) = (n1[0], n1[1]);
            for l in 1..(1 + min(3, self.nodes.len() - iminus - 2)) {
                let (enode, esucc) = (self.nodes[iminus+l],
                                      self.nodes[iminus+l+1]);
                for (p, n2) in self.nodes.windows(2).enumerate().take(iminus){
                    let (pnode, psucc) = (n2[0], n2[1]);
                    let delta = self.or_delta(data, ipred, inode, enode, esucc,
                                              pnode, psucc);
                    if delta < 0 {
			// first copy sequence we want to move
                        let mut t = vec![0; l];
                        t.clone_from_slice(&self.nodes[iminus+1..iminus+l+1]);
			// next move around what will end up being right of it
                        self.nodes.copy_within(p+1..iminus+1, p+1+l);
			// finally write in the sequence being moved
                        self.nodes[p+1..p+l+1].copy_from_slice(&t);
                        return true;
                    }
                }
                for (p, n2) in self.nodes.windows(2).enumerate().skip(iminus+1+l){
                    let (pnode, psucc) = (n2[0], n2[1]);
                    let delta = self.or_delta(data, ipred, inode, enode, esucc,
                                              pnode, psucc);
                    if delta < 0 {
			// first copy sequence we want to move
                        let mut t = vec![0; l];
                        t.clone_from_slice(&self.nodes[iminus+1..iminus+l+1]);
			// next move around what will end up being left of it
                        self.nodes.copy_within(iminus+1+l..p+1, iminus+1);
                        // finally write in the sequence being moved
                        self.nodes[p-l+1..p+1].copy_from_slice(&t);
                        return true;
                    }
                }
            }
        }
        return false;
    }

    pub fn lns(&mut self, data: &TSPData, niter: i32) -> i32 {
        let mut checksum: i32 = 0;
        for _iter in 1..(niter+1) {
            // step 0: copy incumbent
            let mut tmp = self.nodes.clone();
            let mut unplanned = Vec::new();
            // step 1: destroy
            let mut t = 1;
            while t < tmp.len() -1 {
                unplanned.push(tmp[t]);
                tmp.remove(t);
                t += 1;
            }
            // step 2: repair
            while unplanned.len() > 0 {
                let mut bestfrom: usize = 0;
                let mut bestto: usize = 0;
                let mut bestcost: i32 = i32::MAX;
                for (fro, k) in unplanned.iter().enumerate() {
                    for (to, ij) in tmp.windows(2).enumerate() {
                        let (i, j) = (ij[0], ij[1]);
                        let delta = data.d(i, *k) + data.d(*k, j) - data.d(i, j);
                        if delta < bestcost {
                            bestcost = delta;
                            bestfrom = fro;
                            bestto = to;
                        }
                    }
                }
		// perform best found insertion
		tmp.insert(bestto + 1, unplanned[bestfrom]);
		unplanned.remove(bestfrom);
		checksum += bestcost;
            }
            // step 3: set new incumbent
            self.nodes = tmp;
        }
        return checksum;
    }

    pub fn espprc(&self, data: &mut TSPData,
                  nresources: usize, resourcecapacity: usize) -> i32 {
        let mut dual = vec![0; data.n];
        for ij in self.nodes.windows(2) {
            let (i, j) = (ij[0], ij[1]);
            dual[j] = data.d(i, j);
        }
        for i in 0..data.n {
            for j in 0..data.n {
                data.setaux(i, j, f64::from(data.d(i, j) - dual[j]));
            }
        }
        // max len: sum over best assignment for each node
        let mut maxlen: i32 = 0;
        for i in 0..data.n {
            let mut best = i32::MAX;
            for j in 0..data.n {
                if i == j { continue }
                if data.d(i, j) < best {
                    best = data.d(i, j);
                }
            }
            maxlen += best;
        }
        return espprc::solve(data, nresources, resourcecapacity, maxlen);
    }
}

File: implementations/julia-array/TSPBenchmark/src/tsp.jl
struct TSPData{ T<:Real }
    n::Int    # number of nodes
    d::Matrix{T} # distance matrix
    aux::Matrix{Float64}  # auxiliary graph used e.g. in ESPPRC
    aux2::Matrix{Float64}  # auxiliary graph used e.g. in maxflow
end

include("espprc.jl")
include("espprc-index.jl")
include("maxflow.jl")

mutable struct TSPSolution
    tour::Vector{Int}
end

function read_data(fname::String, T::DataType=Int)
    lines = open(fname) do f
        collect(eachline(f))
    end
    n = 0
    drow = 1
    d = zeros(T, 1, 1)
    nsols = 0
    solutions = TSPSolution[]
    for line in lines
        line = replace(line, r"#.*" => "")
        tokens = split(line)
        if length(tokens) == 0
            continue
        elseif length(tokens) == 2 && n == 0
            n = parse(Int, tokens[1])
            nsols = parse(Int, tokens[2])
            d = zeros(T, n, n)
        elseif drow ≤ n
            row = [parse(T, x) for x in tokens]
            d[drow,:] = row
            drow += 1
        elseif length(solutions) ≤ nsols
            push!(solutions, TSPSolution([1+parse(Int, x) for x in tokens]))
        end
    end
    aux = zeros(Float64, size(d))
    aux2 = zeros(Float64, size(d))
    data = TSPData{T}(n, d, aux, aux2)
    return data, solutions
end

function two_opt(data::TSPData{T}, sol::TSPSolution) where T
    t = 0
    while first2eimprovement(data, sol)
        t += 1
    end
    return t
end

function first2eimprovement(d::TSPData{T}, sol::TSPSolution) where T
    for p1 in 1:length(sol.tour)-3, p2 in p1+2:length(sol.tour)-1
        if d.d[sol.tour[p1], sol.tour[p1+1]] +
            d.d[sol.tour[p2], sol.tour[p2+1]] >
            d.d[sol.tour[p1], sol.tour[p2]] +
            d.d[sol.tour[p1+1], sol.tour[p2+1]]
            # improving 2-exchange found
            for i in 0:(p2-p1+1) ÷ 2 - 1
                sol.tour[p1+1+i], sol.tour[p2-i] =
                    sol.tour[p2-i], sol.tour[p1+1+i]
            end
            return true
        end
    end
    return false
end

function or_opt(data::TSPData{T}, sol::TSPSolution) where T
    t = 0
    while firstorimprovement(data, sol)
        t += 1
    end
    return t
end

function firstorimprovement(d::TSPData{T}, s::TSPSolution) where T
    for i in 2:length(s.tour)-1, l in 1:(min(3, length(s.tour)-i)),
        p in [ 1:i-2 ; i+l:length(s.tour)-1 ]
        #
        δ = d.d[s.tour[i-1], s.tour[i+l]] + d.d[s.tour[p], s.tour[i]] +
            d.d[s.tour[i+l-1], s.tour[p+1]] -
            d.d[s.tour[p], s.tour[p+1]] - d.d[s.tour[i-1], s.tour[i]] -
            d.d[s.tour[i+l-1], s.tour[i+l]]
        # improvement found
        if δ < 0
            seq = [ k for k in s.tour[i:i+l-1] ]
            if i < p
                for j in i:p-l
                    s.tour[j] = s.tour[j+l]
                end
                for j in 1:l
                    s.tour[p+j-l] = seq[j]
                end
            else
                for j in i-1:-1:p+1
                    s.tour[j+l] = s.tour[j]
                end
                for j in 1:l
                    s.tour[p+j] = seq[j]
                end
            end
            return true
        end
    end
    return false
end

function lns(d::TSPData{T}, sol::TSPSolution, niter::Int=10) where T
    checksum = zero(T)
    for iter in 1:niter
        # step 0: copy incumbent
        tmp = copy(sol.tour)
        unplanned = Int[]
        # step 1: destroy
        t = 2
        while t < length(tmp)
            push!(unplanned, tmp[t])
            deleteat!(tmp, t)
            t += 1
        end
        # step 2: repair
        while length(unplanned) > 0
            bestcost, bestfrom, bestto = typemax(T), 0, 0
            for from in 1:length(unplanned), to in 1:length(tmp)-1
                δ = d.d[tmp[to], unplanned[from]] +
                    d.d[unplanned[from], tmp[to+1]] -
                    d.d[tmp[to], tmp[to+1]]
                if δ < bestcost
                    bestcost, bestfrom, bestto = δ, from, to
                end
            end
            # perform best found insertion
            insert!(tmp, bestto + 1, unplanned[bestfrom])
            deleteat!(unplanned, bestfrom)
            checksum += bestcost
        end
        # step 3: move to new incumbent
        sol.tour = tmp
    end
    checksum
end

function espprc(d::T, sol::ST;
                nresources::Int=6, resourcecapacity::Int=1,
                index::Bool=false) where T <: TSPData where ST <: TSPSolution
    dual = zeros(Float64, d.n)
    for (i, j) in @views zip(sol.tour[1:end-1], sol.tour[2:end])
        dual[j] = d.d[i, j]
    end
    for i ∈ 1:d.n, j ∈ 1:d.n
        d.aux[i, j] = d.d[i, j] - dual[j]
    end
    # max len: sum of best assignments
    maxlen = sum( [ minimum([d.d[i, j] for j ∈ 1:d.n if i ≠ j])
                    for i ∈ 1:d.n ] )
    e = ESPPRC(d, nresources, resourcecapacity, maxlen)
    if index
        Int(trunc(solvewithindex(e)))
    else
        Int(trunc(solve(e)))
    end
end

function maxflow(d::T, sol::ST) where T <: TSPData where ST <: TSPSolution
    t = zeros(Float64, d.n)
    for (i, j) ∈ @views zip(sol.tour[1:end-1], sol.tour[2:end])
        t[j] = d.d[i, j]
    end
    for i ∈ 1:d.n, j ∈ 1:d.n
        d.aux[i, j] = if (d.d[i, j] > t[j]) d.d[i,j]/1000 else zero(Float64) end
    end
    checksum = zero(Float64);
    for sink in 2:d.n
        mf = edmondskarp(d.aux, d.aux2, d.n, 1, sink)
        checksum += mf
    end
    Int(trunc(checksum))
end
